astropy/table/table.py

```diff
@@ -3274,6 +3274,19 @@ def insert_row(self, index, vals=None, mask=None):
         else:
             raise TypeError("Vals must be an iterable or mapping or None")
 
+        if N == 0:
+            # see https://github.com/astropy/astropy/issues/15964
+            dtype = []
+            for dt, _ in self.dtype.fields.values():
```

Comment:
Actually, seeing this again, I think we should not be getting the dtype: if a user set explicit dtypes, then presumably they knew what they were doing and we should not just override, but do `insert_row` as usual. I'd suggest instead to do:
```
        # If we have effectively an uninitialized table, we feel free to
        # just replace the temporary columns with our row.
        # see https://github.com/astropy/astropy/issues/15964
        if N == 0 and all(col.dtype == float for col in self.columns.values()):
            self.__init__(rows=[vals], names=self.colnames)
            return
```
Now the main question is whether this is a good-enough check that a user never initialized any columns (one could also check `col.dtype == float and col.description is None and col.meta == {}`...). One option would be to mark the columns as uninitialized. For instance, at l.849 in the initialization for `data is None`, we could replace `data = [[]] * n_cols` with something like 
```
data = [Column(description="__uninitialized__")] * n_cols
```
and then do
```
        if N == 0 and all(
            col.description == "__uninitialized__" for col in self.columns.values()
        ):
```
What do you think?

Reply:
This seems reasonable and appears to pass all tests locally, though this implies that units will still be dropped in the following scenario:
```python
from astropy.table import QTable
from astropy import units as u
qt = QTable(names=["a"], dtype="i")
qt.add_row([1*u.cm])
```
And I tend to agree with you that if `dtype` was specified but no value was provided, it seems like the right thing to do (it also neatly provides a way to explicitly opt-in backward compatible behaviour !), but before I make it a test case let's make double sure we agree on this point.

Regarding your proposed marker description for uninitialised columns, I think it's a great idea to make that special case as explicit as possible, and your implementation is sound, so: yoink.